### PR TITLE
RFC32: CKB VM version selection

### PR DESCRIPTION
**Pull request by doitian**

### CODE REVIEW

**Comment by doitian** [comment]
@yangby-cryptape https://github.com/nervosnetwork/rfcs/pull/238/commits/6f113856b997de6188091d1650fd9b2f71cf4c69

File: rfcs/0232-ckb-vm-version-selection/0232-ckb-vm-version-selection.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+## Specification
+
+The CKB VM upgrades via hard fork. The next scheduled hard fork is ckb2021, which is activated since a specific epoch. For all the transactions in the blocks before the activation epoch, they must run the CKB VM version 0 to verify all the script groups. In these transactions, the `hash_type` in cell lock and type script must be either 0 or 1 in the serialized molecule data.
+
+After ckb2021 is activated, CKB node must choose the CKB VM version for each script group. The allowed values for the `hash_type` field in the lock and type script are 0, 1, and 2. Cells are sorted into different groups if they have different `hash_type`. According to the value of `hash_type`:
+
+* When the `hash_type` is 0, the script group matches code via data hash and will run the code using the CKB VM version 0.
+* When the `hash_type` is 1, the script group matches code via type script hash and will run the code using the CKB VM version 1.
+* When the `hash_type` is 2, the script group matches code via data hash and will run the code using the CKB VM version 2.
```
Comment (yangby-cryptape) [inline rfcs/0232-ckb-vm-version-selection/0232-ckb-vm-version-selection.md:30]:
```suggestion
* When the `hash_type` is 2, the script group matches code via data hash and will run the code using the CKB VM version 1.
```

**Review by yangby-cryptape — CHANGES_REQUESTED** [review]

File: rfcs/0032-ckb-vm-version-selection/0032-ckb-vm-version-selection.md [hunk 2]
```diff
@@ -1 +1,81 @@
...
+* 2: "data1"
+
+### CKB VM Versions
+
+The VM version 0 deployed with the Lina and Aggron genesis block is version 0.
+
+The VM version 1 will be deployed along ckb2021. The [rfc33] highlights the changes in this version, and [rfc34] lists the new syscalls only available since version 1.
+
+[rfc33]: ../0033-ckb-vm-version-1/0033-ckb-vm-version-1.md
+[rfc34]: ../0034-vm-syscalls-2/0034-vm-syscalls-2.md
```
Comment (yangby-cryptape) [inline rfcs/0032-ckb-vm-version-selection/0032-ckb-vm-version-selection.md:66]: RFC-0035 should be listed here, too.
Those changes are all due to this RFC.

RFC-0032, RFC-0033, RFC-0034 and RFC-0035 have to be enabled together.

File: rfcs/0032-ckb-vm-version-selection/0032-ckb-vm-version-selection.md [hunk 3]
```diff
@@ -1 +1,81 @@
...
+---
+Number: "0032"
+Category: Consensus (Hard Fork)
+Status: Draft
+Author: Ian Yang
+Organization: Nervos Foundation
+Created: 2021-04-26
+---
+
+# CMB VM Version Selection
```
Comment (yangby-cryptape) [inline rfcs/0032-ckb-vm-version-selection/0032-ckb-vm-version-selection.md:10]:
```suggestion
# CKB VM Version Selection
```

**Review by yangby-cryptape — CHANGES_REQUESTED** [review]

File: rfcs/0032-ckb-vm-version-selection/0032-ckb-vm-version-selection.md [hunk 4]
```diff
@@ -1 +1,81 @@
...
+* Before the fork is activated, CKB node must relay transactions via relay protocol C and must drop all messages received via protocol N.
+* After the fork is activated, CKB node must relay transactions via relay protocol N and must drop all messages received via protocol C.
+
+The relay protocol C will be dropped after the fork succeeds. See [rfc35] for details.
+
+[rfc35]: ../0035-ckb2021-p2p-protocol-upgrade/0035-ckb2021-p2p-protocol-upgrade.md
+
+When a new block is appended to the chain and the fork is activated, or a block is rolled back and the fork is deactivated, the CKB node must rerun the verification on all the transactions in the pool.
+
+In the [Rust implementation of CKB](https://github.com/nervosnetwork/ckb) provided by Nervos Foundation, the `hash_type` is returned as an enum. Now it has three allowed values:
```
Comment (doitian) [inline rfcs/0032-ckb-vm-version-selection/0032-ckb-vm-version-selection.md:53]:
```suggestion
In [nervosnetwork/ckb](https://github.com/nervosnetwork/ckb), the `hash_type` is returned in the JSON RPC as an enum. Now it has three allowed values:
```

File: rfcs/0032-ckb-vm-version-selection/0032-ckb-vm-version-selection.md [hunk 5]
```diff
@@ -1 +1,81 @@
...
+
+It's essential to keep improving CKB VM because it is the computation bottleneck of the whole network. The upgrade packages can improve the performance, bring bug fixings and add new RISC-V extensions. However the upgrade should not break the old code, users must have the opt-in option to specify the VM version.
+
+This RFC proposes a general mechanism that determines how the CKB node chooses the CKB VM version for a transaction script group.
+
+## Specification
+
+The CKB VM upgrades via hard fork. The next scheduled hard fork is ckb2021, which is activated since a specific epoch. For all the transactions in the blocks before the activation epoch, they must run the CKB VM version 0 to verify all the script groups. In these transactions, the `hash_type` in cell lock and type script must be either 0 or 1 in the serialized molecule data.
+
+After ckb2021 is activated, CKB nodes must choose the CKB VM version for each script group. The allowed values for the `hash_type` field in the lock and type script are 0, 1, and 2. Cells are sorted into different groups if they have different `hash_type`. According to the value of `hash_type`:
```
Comment (doitian) [inline rfcs/0032-ckb-vm-version-selection/0032-ckb-vm-version-selection.md:26]: TODO: add cross reference between this RFC and the transaction structure RFC.

File: rfcs/0032-ckb-vm-version-selection/0032-ckb-vm-version-selection.md (hunk 2 above)
Comment (doitian) [reply rfcs/0032-ckb-vm-version-selection/0032-ckb-vm-version-selection.md]: Resolved by dd5f26a02d3db1659f86c42bdbf52c49975a68bb

File: rfcs/0032-ckb-vm-version-selection/0032-ckb-vm-version-selection.md (hunk 5 above)
Comment (doitian) [reply rfcs/0032-ckb-vm-version-selection/0032-ckb-vm-version-selection.md:26]: Resolved by dd5f26a02d3db1659f86c42bdbf52c49975a68bb

File: rfcs/0022-transaction-structure/0022-transaction-structure.md [hunk 6]
```diff
@@ -128,10 +128,14 @@ If you want to use a script in CKB, follow the code locating rules:
 
 - Compile your code into RISC-V binary. You can find some examples in the [repository](https://github.com/nervosnetwork/ckb-system-scripts) which builds the code for system cells.
 - Create a cell which stores the binary as data in a transaction, and send the transaction to the chain.
-- Construct a script structure, which `hash_type` is "Data", and `code_hash` is just the hash of the built binary.
+- Construct a script structure, which `hash_type` is "Data"[^1], and `code_hash` is just the hash of the built binary.
 - Use the script as the type or the lock script in a cell.
 - If the script has to run in a transaction, include the code cell's out point in the `cell_deps`.
 
+[^1]: Or "Data1" after [rfc32] is activated.
```
Comment (xxuejie) [inline rfcs/0022-transaction-structure/0022-transaction-structure.md:135]: This reminds me of something I've long wanted to ask: how do we do proper references in RFC? Should we just inline the link, or use a more traditional reference style footnotes at the end of the document?

Personally I feel using inline links work better since we are writing markdowns anyway, there are not so many chances where one wants to print the RFCs. But this is clearly bigger than one's preference, so the real question is: is there a decided style on references?

File: rfcs/0032-ckb-vm-version-selection/0032-ckb-vm-version-selection.md [hunk 7]
```diff
@@ -1 +1,157 @@
...
+Author: Ian Yang
+Organization: Nervos Foundation
+Created: 2021-04-26
+---
+
+# CKB VM Version Selection
+
+## Abstract
+
+This RFC proposes a mechanism to choose the CKB VM version after the hard fork. What's included in the new CKB VM version is beyond the scope of this document.
```
Comment (xxuejie) [inline rfcs/0032-ckb-vm-version-selection/0032-ckb-vm-version-selection.md:14]: Recommend adding a link to the VM version 1 RFC as an example of a VM version

File: rfcs/0032-ckb-vm-version-selection/0032-ckb-vm-version-selection.md [hunk 8]
```diff
@@ -1 +1,157 @@
...
+
+## Motivation
+
+It's essential to keep improving CKB VM because it is the computation bottleneck of the whole network. The upgrade packages can improve the performance, bring bug fixings and add new RISC-V extensions. However the upgrade should not break the old code, users must have the opt-in option to specify the VM version.
+
+This RFC proposes a general mechanism that determines how the CKB node chooses the CKB VM version for a transaction script group.
+
+## Specification
+
+The CKB VM upgrades via hard fork. The next scheduled hard fork is ckb2021, which is activated since a specific epoch. For all the transactions in the blocks before the activation epoch, they must run the CKB VM version 0 to verify all the script groups. In these transactions, the `hash_type` in cell lock and type script must be either 0 or 1 in the serialized molecule data.
```
Comment (xxuejie) [inline rfcs/0032-ckb-vm-version-selection/0032-ckb-vm-version-selection.md:24]: There is a mention of `CKB VM version 0` out of nowhere, I do recommend we put up a short section at the start of this specification, mentioning that the `Lina` fork version uses CKB VM version 0, while the new ckb2021 fork introduces a new CKB VM version 1.

Comment (xxuejie) [reply rfcs/0032-ckb-vm-version-selection/0032-ckb-vm-version-selection.md:24]: Also it's worth mentioning that the selection of CKB VM versions, is enabled via `hash_type`. Otherwise one can get confused why this paragraph mentions `hash_type` at all, which seems to be completely irrelavant to CKB VM versions.

File: rfcs/0032-ckb-vm-version-selection/0032-ckb-vm-version-selection.md [hunk 9]
```diff
@@ -1 +1,157 @@
...
+
+After ckb2021 is activated, CKB nodes must choose the CKB VM version for each script group. The allowed values for the `hash_type` field in the lock and type script are 0, 1, and 2. Cells are sorted into different groups if they have different `hash_type`. According to the value of `hash_type`:
+
+* When the `hash_type` is 0, the script group matches code via data hash and will run the code using the CKB VM version 0.
+* When the `hash_type` is 1, the script group matches code via type script hash and will run the code using the CKB VM version 1.
+* When the `hash_type` is 2, the script group matches code via data hash and will run the code using the CKB VM version 1.
+
+| `hash_type` | matches by       | VM version |
```
Comment (xxuejie) [inline rfcs/0032-ckb-vm-version-selection/0032-ckb-vm-version-selection.md:32]: I would recommend a short section breaking down the bits of `hash_type`. People will wonder why `0` in hash type means `data hash + VM version 0`, but `2` in hash type means `data hash + VM version 1`(why not use 1 here).

File: rfcs/0032-ckb-vm-version-selection/0032-ckb-vm-version-selection.md [hunk 10]
```diff
@@ -1 +1,157 @@
...
+[rfc33]: ../0033-ckb-vm-version-1/0033-ckb-vm-version-1.md
+[rfc34]: ../0034-vm-syscalls-2/0034-vm-syscalls-2.md
+
+The transaction is considered invalid if any `hash_type` is not in the allowed values 0, 1, and 2.
+
+See more information about code locating using `hash_type` in [rfc22].
+
+[rfc22]: ../0022-transaction-structure/0022-transaction-structure.md
+
+Because the VM selection algorithm depends on which epoch the transaction belongs to, it is not deterministic for transactions still in the memory pool. The CKB node must run two versions of transaction relay protocols, one for the current CKB version C and another for the next fork version N.
```
Comment (xxuejie) [inline rfcs/0032-ckb-vm-version-selection/0032-ckb-vm-version-selection.md:51]: I feel like this is implementation details. We could add this in an appendix, but not here in the specification.

File: rfcs/0032-ckb-vm-version-selection/0032-ckb-vm-version-selection.md [hunk 11]
```diff
@@ -1 +1,157 @@
...
+
+* Before the fork is activated, CKB node must relay transactions via relay protocol C and must drop all messages received via protocol N.
+* After the fork is activated, CKB node must relay transactions via relay protocol N and must drop all messages received via protocol C.
+
+The relay protocol C will be dropped after the fork succeeds. See [rfc35] for details.
+
+[rfc35]: ../0035-ckb2021-p2p-protocol-upgrade/0035-ckb2021-p2p-protocol-upgrade.md
+
+Attention that, [rfc33], [rfc34], and [rfc35] must be activated together with this one starting at the same epoch.
```
Comment (xxuejie) [inline rfcs/0032-ckb-vm-version-selection/0032-ckb-vm-version-selection.md:60]: Maybe rephrase this as dependencies among RFCs?

File: rfcs/0032-ckb-vm-version-selection/0032-ckb-vm-version-selection.md (hunk 7 above)
Comment (doitian) [reply rfcs/0032-ckb-vm-version-selection/0032-ckb-vm-version-selection.md:14]:
```suggestion
This RFC proposes a mechanism to choose the CKB VM version after the hard fork. This document will not cover what's included in the new CKB VM version, please refer to [rfc33] and [rfc34] for the changes.
```

**Comment by doitian** [comment]
Assigned reviewers: @xxuejie @zhangsoledad

File: rfcs/0032-ckb-vm-version-selection/0032-ckb-vm-version-selection.md [hunk 12]
```diff
@@ -1 +1,157 @@
...
+
+* 0: "data"
+* 1: "type"
+* 2: "data1"
+
+## Rationale
+
+There are many other solutions to select VM versions. The current solution is the result of discussion and trade-off. Following are some example alternatives:
+
+* Always use the latest VM version. The users have no options to freeze the VM versions used in their transactions.
```
Comment (zhangsoledad) [inline rfcs/0032-ckb-vm-version-selection/0032-ckb-vm-version-selection.md:74]:
```suggestion
Consistently uses the latest VM version. The users cannot specify the VM versions for transactions, and the version selection will be non-determine cause it will depend on the chain state.
```

File: rfcs/0032-ckb-vm-version-selection/0032-ckb-vm-version-selection.md [hunk 13]
```diff
@@ -1 +1,157 @@
...
+* Depend on the script code cell epoch. Use the old VM version if the code cell is deployed before the fork, and use the new one otherwise. The problem of this solution is that anyone can re-deploy the cell and construct the transaction using the new code cell to choose VM versions.
```
Comment (zhangsoledad) [inline rfcs/0032-ckb-vm-version-selection/0032-ckb-vm-version-selection.md:75]:
```suggestion
Depend on the script code cell deployed epoch. Use the old VM version if the code cell is deployed before the fork, and use the new one otherwise. The drawback of this proposal is that anyone can re-deploy the script code after the fork, then the transaction using the newly deployed code cell will choose new VM versions. This may be unexpected behavior.
```

File: rfcs/0032-ckb-vm-version-selection/0032-ckb-vm-version-selection.md (hunk 9 above)
Comment (doitian) [reply rfcs/0032-ckb-vm-version-selection/0032-ckb-vm-version-selection.md:32]: I'll explain why we choose such encoding pattern.

However, I will not mention the bits level encoding of `hash_type` because:

1. We have not decided yet how to encode `data hash + VM version 2` and following versions
2. We want developers use this field as an enum, instead of a continuous version number plus a type flag.

**Review by xxuejie — APPROVED** [review]

**Review by zhangsoledad — APPROVED** [review]

**Comment by doitian** [comment]
FCP will end on 2022-04-01
